### PR TITLE
Mods for relocation 1

### PR DESCRIPTION
**Pull request by TonyB9000**
@tomvothecoder @chengzhuzhang Whole lotta changes.  Extensive replacement of hardcoded "/p/user_pub" paths with fully relocatable paths (probably 80+% of changes).   Ancillary tools consolidations and additions. Much "comment and log-message" cleanup.

### CODE REVIEW

**Comment by tomvothecoder** [comment]
Hey Tony, thanks for this PR. Can you summarize the PR's objective, high-level changes, what we should focus on during our review process? This help Jill and I review more effectively in our very limited time.

If you need an example, here's one I did for this large e3sm_diags PR: https://github.com/E3SM-Project/e3sm_diags/pull/677. Thanks!

**Comment by TonyB9000** [comment]
**Description**

    Closes no issues, as no issues were raised.

This PR encompasses a variety of unnamed issues:

**(1) Relocatable Paths:** The replacement of ~100 instanced of hard-coded paths in both datasm-proper (python) codes and scores of support scripts (python an bash) was conducted in support of the datasm "**Relocation/Migration**" plan.  These are identified by seeking the term "DSM_GETPATH" throughout these codes, generally taking the form - for bash shell scripts:

```
       <path_var>=`$DSM_GETPATH <path_key>`
```
and for python codes:
```
	import os
	import subprocess

	gp = os.environ['DSM_GETPATH']
	<path_var> = subprocess.run([gp, <path_key>],stdout=subprocess.PIPE,text=True).stdout.strip()
```

**(2) Added Tools:** Many subordinate operations scripts added to the tools directory were also added for the first time.  Each should contain explanatory help text at the top, detailing their function.

**(3) Fixes to CMIP6 Generation and Publication:** Several changes to the datasm "task transition" handling, and related changes and fixes discovered in the course of CMIP6 generation and publication of the Large Ensemble datasets are included, as some of these overlap with the path-relocation changes.  The term "dataset_spec" should accompany the changes made to allow alternative dataset_spec files to be supplied on the command line, as we do not include "external request" publications in our E3SM dataset_spec.yaml file.

**(4) Consolidation Support:** Several of the "Generate*CMIP6.py" modules were restructured to force as much line-by-line similarity as possible, in anticipation of a future consolidation of these codes.

**(5) Logs and Comments:** Changes to comments and "log_message" additions are plentiful, and can be glossed over.

**Comment by TonyB9000** [comment]
@tomvothecoder (Your formatting skills are far superior to my own...)

File: datasm/dataset.py [hunk 1]
```diff
@@ -94,7 +94,7 @@ def __init__(
 
         self.versions = versions
 
-        log_message("debug", f"init: splitting to obtain facets: dsid = {self.dataset_id}")
+        log_message("debug", f"dataset.py init: splitting to obtain facets: dsid = {self.dataset_id}")
```
Comment (tomvothecoder) [inline datasm/dataset.py:97]:
```suggestion
        log_message("debug", f"{__name__} init: splitting to obtain facets: dsid = {self.dataset_id}")
```
You can use `__name__` to get the name of the module instead of manually typing it out

File: datasm/dataset.py [hunk 2]
```diff
@@ -129,6 +127,7 @@ def __init__(
                     break
             if self.table == "fx" or self.table == "Ofx":
                 self.freq = "fixed"
+                # self.freq = "mon"
```
Comment (tomvothecoder) [inline datasm/dataset.py:130]:
```suggestion
```

Deleting commented out code.

File: datasm/scripts/publish_to_esgf.py [hunk 3]
```diff
@@ -112,7 +113,8 @@ def publish_dataset(args):
             )
             proc.wait()
 
-        con_message("info", f"Return code {str(proc.returncode)} on cmd: {cmd}")      # generated weird error in 3_Publish/slurm_scripts-20211013_185431_334581
+        con_message("info", f"Return code {str(proc.returncode)} on cmd: {cmd}")
+        log_message("info", f"Return code {str(proc.returncode)} on cmd: {cmd}")
```
Comment (tomvothecoder) [inline datasm/scripts/publish_to_esgf.py:117]: Can we just keep the line for `log_message()`?

File: datasm/tools/run_cwltool_test.sh [hunk 4]
```diff
@@ -6,12 +6,13 @@ in_cmip_dsid=$1
 dry_run=1
 
 # external dependencie
-get_cmip6_parms=/p/user_pub/e3sm/staging/tools/cmip6_parameters.sh
-force_status=/p/user_pub/e3sm/staging/tools/ensure_status_file_for_dsid.sh
-list_e3sm_dsids=/p/user_pub/e3sm/staging/tools/list_e3sm_dsids.py
-meta_version=/p/user_pub/e3sm/staging/tools/metadata_version.py
+dsm_tools=`$DSM_GETPATH STAGING_TOOLS`
+resource_path=`$DSM_GETPATH STAGING_RESOURCE`
```
Comment (tomvothecoder) [inline datasm/tools/run_cwltool_test.sh:10]: I noticed similar paths get repeated in other tools. 

Is it possible to consolidate these paths into a single file then reference them based on the needs of each file?
This will cute down on repeated code and will probably make maintaining the tools easier.

File: datasm/tools/run_datasm_LocalEnv_postprocess_dsid_list_serially.sh [hunk 5]
```diff
@@ -0,0 +1,93 @@
+#!/bin/bash
+
+usage="run_datasm_postprocess_dsid_list_serially.sh file_of_dataset_ids [alt_path_to_dataset_spec]"
+
+# NOTE: this version relies upon [STAGING_TOOLS]/latest_data_location_by_dsid.sh to select the src_root
+
+# obtain global relocatable paths
+
+dsm_tools=`$DSM_GETPATH STAGING_TOOLS`
+resource_path=`$DSM_GETPATH STAGING_RESOURCE`
```
Comment (tomvothecoder) [inline datasm/tools/run_datasm_LocalEnv_postprocess_dsid_list_serially.sh:10]: Example of instantiating same paths again in another tool

File: datasm/tools/run_esgunpublish.sh [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    if [ $2 == "DELETE" ]; then
+        do_delete="--delete"
+        target_op="Deleted"
+    fi
+fi
+
+resource_path=`$DSM_GETPATH STAGING_RESOURCE`
+dsm_tools=`$DSM_GETPATH STAGING_TOOLS`
+pub_root=`$DSM_GETPATH PUBLICATION_DATA`
+statpath=`$DSM_GETPATH STAGING_STATUS`
```
Comment (tomvothecoder) [inline datasm/tools/run_esgunpublish.sh:26]: Example of instantiating the same paths again

File: datasm/tools/tell_years_dsid.py [hunk 7]
```diff
@@ -23,7 +23,8 @@ def assess_args():
 
     return args
 
-resource_path = '/p/user_pub/e3sm/staging/resource/'
+gp = os.environ['DSM_GETPATH']
```
Comment (tomvothecoder) [inline datasm/tools/tell_years_dsid.py:26]: I suggest defining `DSM_GETPATH = os.environ["DSM_GETPATH"]` in a single file under `datasm/tools` and importing them in the Python files that need it. If you ever rename `DSM_GETPATH` (even if it is unlikely), you will need to find all instances of it across the Python tools and replace it.

`env.py`:
```python
import os

# DEFINE ENV VARS HERE
DSM_GETPATH = os.environ["DSM_GETPATH"]
```

`tell_years_dsid.py`:
```python
...
from datasm.tools.env import DSM_GETPATH
...
```

File: datasm/workflows/jobs/__init__.py [hunk 8]
```diff
@@ -236,11 +244,18 @@ def requires_dataset(self, dataset):
...
-
             rcode = dataset.realm.replace('-','')
             gcode = dataset.grid.replace('-','')
             fcode = dataset.freq.replace('-','')
+
+            self_rcode = self.dataset.realm
+            self_fcode = self.dataset.freq
+
+            # log_message("info", f"init: (rcode,fcode) = ({rcode},{fcode})")
+            # log_message("info", f"init: (self_rcode,self_fcode) = ({self_rcode},{self_fcode})")
```
Comment (tomvothecoder) [inline datasm/workflows/jobs/__init__.py:255]:
```suggestion
```

Deleting commented out code.

File: datasm/resources/datasm_config.yaml [hunk 9]
```diff
@@ -1,3 +1,6 @@
+### WARNING - DO NOT EDIT DIRECTLY - changes can be overwritten.    ###
+### USE (resource)/datasm_config.template.sh to modity and recreate ###
```
Comment (tomvothecoder) [inline datasm/resources/datasm_config.yaml:2]: This file gets edited through `(resource)/datasm_config.template.sh`?

**Review by tomvothecoder — APPROVED** [review]
Hi Tony, I scanned your changes and left some minor comments and questions. Feel free to address/not address and merge.

**Comment by TonyB9000** [comment]
> Hi Tony, I scanned your changes and left some minor comments and questions. Feel free to address/not address and merge.

@tomvothecoder Thanks for the very detailed and exhaustive review!  I will attempt to implement many of these suggestions.

I want to discuss some of the "relocatable paths" issues with your advice in a separate email, where I can feel free to elaborate.  Will capture salient issues here, thereafter.

File: datasm/dataset.py (hunk 1 above)
Comment (TonyB9000) [reply datasm/dataset.py:97]: @tomvothecoder Excellent.  I'll seek to replace these.

Is there a corresponding `{__func__}` that will obtain the name of the function in which a message is generated?

File: datasm/scripts/publish_to_esgf.py (hunk 3 above)
Comment (TonyB9000) [reply datasm/scripts/publish_to_esgf.py:117]: Yes.  Sterling was real big on flashy/colorful console output inappropriate for batch-mode operation, where all such output gets redirected to log-files anyway.  Post processing of such logs (sorting messages by timestamp) would fail because the leading bytes of message lines are color-codes...

File: datasm/tools/tell_years_dsid.py (hunk 7 above)
Comment (TonyB9000) [reply datasm/tools/tell_years_dsid.py:26]: I will study this.  The idea is to allow operators to redefine these "datasm system paths" without having to rebuild or re-install datasm code.  The "DSM_GETPATH" (for python modules) calls an external executable.  We could hide that in a single import module I guess.

So, instead of each module doing:

```
gp = os.environ['DSM_GETPATH']
wh_root = subprocess.run([gp, "STAGING_DATA"],stdout=subprocess.PIPE,text=True).stdout.strip()
pb_root = subprocess.run([gp, "PUBLICATION_DATA"],stdout=subprocess.PIPE,text=True).stdout.strip()
(etc etc etc as needed)
```

We obtain ALL "DSM_ROOT_PATHS" at once, and import the resulting array of paths(?)  e.g.:

from datasm.util import set_dsm_paths

```
dsm_paths = dsm_set_paths()
wh_root = dsm_paths['STAGING_DATA']
pb_root = dsm_paths['PUBLICATION_DATA']
(etc etc etc as needed)
```

Or is there an even better option?

File: datasm/resources/datasm_config.yaml (hunk 9 above)
Comment (TonyB9000) [reply datasm/resources/datasm_config.yaml:2]: Yes, it gets "built", because the top-level paths are defined by an external OS call to a file the user can manipulate (path values, not the path keys) without having to build/reinstall datasm.  They would still need to rerun datasm_config.template.sh

Comment (TonyB9000) [reply datasm/resources/datasm_config.yaml:2]: @tomvothecoder In truth, the path-part of that config (at the top) should eventually be removed (they inherit from the external dsm_paths config in the first place) and could then become a static config once again.   Another day.

**Comment by TonyB9000** [comment]
@tomvothecoder I have made most of the changes you had suggested.  The biggest is to have added a function to util.py "get_dsm_paths()" that makes a system call to obtain the current filesystem layout.  It is not "baked in" to datasm.  Once that "dictionary" is produced, it is handed back to the process that needs the path(s), so each process need not call "subprocess", nor does the system call need to be made for each path individually.

So much for the python part.  For the bash scripts, something similar can be done.  But once again, even if the paths are embedded in some bash-file of paths that can simply be "included" into a given script, the values must still be instantiated at runtime, for they are allowed to change independent of the scripts themselves.  So, an ENV variable (in the user's .bashrc) must still exist to indicate the location of the process that will make those paths available.

Let me know if you'd like to re-review the changes.  I'm glad you provided the impetus for me to further consolidate the path management.
